ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1531 @@
-				break;
-			}
-		}
-		mbfl_convert_filter_flush(encoder);
-		if (pc.status != 0 && mkwidth > 0) {
-			pc.width += mkwidth;
-			if (n > 0) {
-				while (n > 0) {
-					if ((*encoder->filter_function)(*p++, encoder) < 0) {
-						break;
-					}
-					n--;
-				}
-				mbfl_convert_filter_flush(encoder);
-			} else if (pc.outwidth > pc.width) {
-				pc.status++;
-			}
-			if (pc.status != 1) {
-				pc.status = 10;
-				pc.device.pos = pc.endpos;
-				mbfl_convert_filter_copy(pc.decoder_backup, pc.decoder);
-				mbfl_convert_filter_reset(encoder, marker->encoding, &mbfl_encoding_wchar);
-				p = marker->val;
-				n = marker->len;
-				while (n > 0) {
-					if ((*encoder->filter_function)(*p++, encoder) < 0) {
-						break;
-					}
-					n--;
-				}
-				mbfl_convert_filter_flush(encoder);
+
+	/* Now check how far we can get into remainder of string while still leaving
+	 * enough space for the trim marker, in case it is necessary to add it */
+	size_t marker_width = (marker ? mbfl_strwidth(marker) : 0);
+
+	size_t width = 0;
+	mbfl_convert_filter *filter = mbfl_convert_filter_new(string->encoding,
+		&mbfl_encoding_wchar, filter_count_width, 0, &width);
+
+	unsigned char *p = start, *enough_space_for_marker = start;
+	while (p < e) {
+		mbfl_filter_feed_char(*p++, filter);
+		if (width + marker_width <= max_width) {
+			/* If `filter->status` is zero, we have just passed over a complete
+			 * character and are entering another one */
+			if (filter->status == 0) {
+				enough_space_for_marker = p;
 			}
-		} else if (pc.status != 0) {
-			pc.device.pos = pc.endpos;
-			mbfl_convert_filter_copy(pc.decoder_backup, pc.decoder);
+		} else {
+			break;
 		}
-		mbfl_convert_filter_flush(pc.decoder);
 	}
-	result = mbfl_memory_device_result(&pc.device, result);
-	mbfl_convert_filter_delete(encoder);
-	mbfl_convert_filter_delete(pc.decoder);
-	mbfl_convert_filter_delete(pc.decoder_backup);
+
+	/* OK, we've found the last position where the trim marker could fit.
+	 * Now let's see if the rest of the string will fit in `max_width` or not */
+	while (width < max_width && p < e) {
+		mbfl_filter_feed_char(*p++, filter);
+	}
+	mbfl_convert_filter_delete(filter);
+
+	if (p == e && width <= max_width) {
+		/* The whole input string fits within `max_width`, at least after skipping
+		 * over `from` characters, so there is no need to add the trim marker */
+		size_t length = string->val + string->len - start;
+		result->val = ecalloc(length, sizeof(unsigned char));
+		memcpy(result->val, start, length);
+		result->len = length;
+	} else {
+		/* We need to trim the input string and add the trim marker */
+		size_t trimmed_length = enough_space_for_marker - start;
+		size_t marker_length  = (marker ? marker->len : 0);
+		result->val = ecalloc(trimmed_length + marker_length, sizeof(unsigned char));
+		memcpy(result->val, start, trimmed_length);
+		memcpy(result->val + trimmed_length, marker->val, marker_length);
```

Comment:
ubsan will complain about that `marker->val`. Wrap the whole memcpy in `if (marker)` to avoid.

Reply:
Though, to offer an alternative here to simplify the code further: A NULL marker and an empty marker are equivalent. We could require a non-null marker and make the caller pass an empty string. This is trivial for our usage (just a matter of replacing a NULL with "").

Reply:
Hmm. Will think about that.

Reply:
After thinking about this, I prefer to allow a NULL marker, but use `if (marker)` to avoid UB.